src/tcn_ext/tcn_keys.rs

```diff
@@ -1,80 +1,169 @@
-      Self::bytes_to_rak(rak_bytes)
-
-    ).unwrap_or({
-      let new_key = ReportAuthorizationKey::new(rand::thread_rng());
-      self.preferences.set_autorization_key(Self::rak_to_bytes(new_key));
-      new_key
-    })
-  }
-
-  fn tck(&self) -> TemporaryContactKey {
-    self.preferences.tck().map(|rak_bytes|
-      Self::bytes_to_tck(rak_bytes)
-
-    ).unwrap_or({
-      self.rak().initial_temporary_contact_key()
-    })
-  }
-
-  fn set_tck(&self, tck: TemporaryContactKey) {
-    self.preferences.set_tck(Self::tck_to_bytes(tck));
-  }
-
-  fn rak_to_bytes(rak: ReportAuthorizationKey) -> [u8; 32] {
-    let mut buf = Vec::new();
-    rak.write(Cursor::new(&mut buf))
-      .expect("Couldn't write RAK bytes");
-    Self::byte_vec_to_32_byte_array(buf)
-  }
-
-  fn byte_vec_to_32_byte_array(bytes: Vec<u8>) -> [u8; 32] {
-    let mut array = [0; 32];
-    let bytes = &bytes[..array.len()]; // panics if not enough data
-    array.copy_from_slice(bytes); 
-    array
-  }
-
-  fn bytes_to_rak(bytes: [u8; 32]) -> ReportAuthorizationKey {
-    ReportAuthorizationKey::read(Cursor::new(&bytes))
-      .expect("Couldn't read RAK bytes")
-  }
-
-  fn tck_to_bytes(tck: TemporaryContactKey) -> [u8; 32] {
-    let mut buf = Vec::new();
-    tck.write(Cursor::new(&mut buf))
-      .expect("Couldn't write TCK bytes");
-    Self::byte_vec_to_32_byte_array(buf)
-  }
-
-  fn bytes_to_tck(bytes: [u8; 32]) -> TemporaryContactKey {
-    TemporaryContactKey::read(Cursor::new(&bytes))
-      .expect("Couldn't read TCK bytes")
-  }
+impl<PreferencesType: Preferences> TcnKeysImpl<PreferencesType> {
+    fn rak(&self) -> ReportAuthorizationKey {
+        self.preferences
+            .authorization_key()
+            .map(|rak_bytes| Self::bytes_to_rak(rak_bytes))
+            .unwrap_or({
+                let new_key = ReportAuthorizationKey::new(rand::thread_rng());
+                self.preferences
+                    .set_autorization_key(Self::rak_to_bytes(new_key));
+                new_key
+            })
+    }
+
+    fn tck(&self) -> TemporaryContactKey {
+        self.preferences
+            .tck()
+            .map(|tck_bytes| Self::bytes_to_tck(tck_bytes))
+            .unwrap_or({ self.rak().initial_temporary_contact_key() })
+    }
+
+    fn set_tck(&self, tck: TemporaryContactKey) {
+        self.preferences.set_tck(Self::tck_to_bytes(tck));
+    }
+
+    fn rak_to_bytes(rak: ReportAuthorizationKey) -> [u8; 32] {
+        let mut buf = Vec::new();
+        rak.write(Cursor::new(&mut buf))
+            .expect("Couldn't write RAK bytes");
+        Self::byte_vec_to_32_byte_array(buf)
+    }
+
+    fn byte_vec_to_32_byte_array(bytes: Vec<u8>) -> [u8; 32] {
+        let mut array = [0; 32];
+        let bytes = &bytes[..array.len()]; // panics if not enough data
+        array.copy_from_slice(bytes);
+        array
+    }
+
+    fn byte_vec_to_tck_byte_array(bytes: Vec<u8>) -> TckBytesWrapper {
```

Comment:
I'd name it `byte_vec_to_tck_byte_array `. This name confused me a bit in the tests below.

Reply:
Agreed.